jvm-libs/linea/web3j-extensions/src/main/kotlin/linea/web3j/SearchCursor.kt

```diff
@@ -0,0 +1,76 @@
+package linea.web3j
+
+import linea.SearchDirection
+
+/**
+ * SearchCursor is a helper class to iterate over a range of ULong values in a binary search manner.
+ * When search is not provided returns the next unsearched chunk without moving left or right cursors because
+ * no direction is provided, but flags the chunk as searched to not go over again.
+ *
+ * This caveat is because when searching for EthLogs on L1, a blockInterval (represend by chunk here) may not have logs
+ * so the search predicate cannot tell what direction to go next, so we need to flag the chunk as searched and try the
+ * next chunk.
+ */
+class SearchCursor(
+  val from: ULong,
+  val to: ULong,
+  val chunkSize: Int
+) {
+  private data class Chunk(val interval: Pair<ULong, ULong>, var searched: Boolean = false)
+
+  private val searchChunks = (from..to)
+    .chunked(chunkSize)
+    .map { Chunk(it.first() to it.last(), searched = false) }
+    .toMutableList()
+  private var left = 0
+  private var right = searchChunks.size - 1
+  private var prevCursor: Int? = null
+
+  @Synchronized
+  fun next(searchDirection: SearchDirection?): Pair<ULong, ULong>? {
+    return if (left > right) {
+      null
+    } else {
+      if (prevCursor == null) {
+        // 1st call, lets start in the middle
+        val mid = left + (right - left) / 2
+        searchChunks[mid] to mid
+      } else {
+        if (searchDirection == null) {
+          findNextUnsearchedChunk()
+        } else {
+          if (searchDirection == SearchDirection.FORWARD) {
+            left = prevCursor!! + 1
+          } else {
+            right = prevCursor!! - 1
+          }
+          if (left > right) {
+            null
+          } else {
+            val mid = left + (right - left) / 2
+            val chunk = searchChunks[mid]
+            if (chunk.searched) {
+              // we have already searched this chunk, lets find next unsearched
+              findNextUnsearchedChunk()
+            } else {
+              chunk to mid
+            }
```

Comment:
I think this whole chunk of code can be reused in the `prevCursor == null` condition. Even though check `chunk.searched` is not needed in the first iteration, reusing this bit may make this function easier to read.